### PR TITLE
Dataframe Indexing interface for `Data`

### PR DESCRIPTION
**Pull request by eliascarv**

### CODE REVIEW

File: src/traits/data.jl [hunk 1]
```diff
@@ -148,6 +148,83 @@ end
...
+  pairs = (nm => Tables.getcolumn(cols, nm)[row] for nm in names)
+  (; pairs..., geometry=dom[row])
+end
+
+function Base.getindex(data::Data, rows::AbstractVector{Int}, names::AbstractVector{Symbol})
+  dom   = domain(data)
+  table = values(data)
+  cols  = Tables.columns(table)
+
+  newdom = Collection(dom[row] for row in rows)
```
Comment (juliohm) [inline src/traits/data.jl]: We should use view(dom, rows) instead

File: src/traits/data.jl [hunk 2]
```diff
@@ -148,6 +148,83 @@ end
...
+  getproperty(data, col)[row]
+
+function Base.getindex(data::Data, rows::AbstractVector{Int}, col::Symbol)
+  if col == :geometry
+    dom = domain(data)
+    Collection(dom[row] for row in rows)
+  else
+    cols = Tables.columns(values(data))
+    column = Tables.getcolumn(cols, col)
+    [column[row] for row in rows]
```
Comment (juliohm) [inline src/traits/data.jl]: Can we use Tables.subset instead?

File: src/traits/data.jl [hunk 3]
```diff
@@ -136,17 +136,124 @@ Tables.materializer(D::Type{<:Data}) = D
...
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, vars::AbstractVector{Symbol})
+  _checkvars(vars)
+  _rmgeometry!(vars)
+  table = values(data)
+
+  newdom = view(domain(data), inds)
+  subset = Tables.subset(table, inds)
+  cols   = Tables.columns(subset)
+  𝒯 = NamedTuple(var => Tables.getcolumn(cols, var) for var in vars)
```
Comment (juliohm) [inline src/traits/data.jl]: let's use the same pattern with the variable `pairs` here as well

File: src/traits/data.jl [hunk 4]
```diff
@@ -136,17 +136,124 @@ Tables.materializer(D::Type{<:Data}) = D
...
+end
+
+Base.getindex(data::Data, inds::AbstractVector{Int}, vars::AbstractVector{<:AbstractString}) =
+  getindex(data, inds, Symbol.(vars))
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, ::Colon)
+  table = values(data)
+
+  newdom   = view(domain(data), inds)
+  newtable = Tables.subset(table, inds)
```
Comment (juliohm) [inline src/traits/data.jl]: In this method I think we can simply use view(data ,inds)

File: src/traits/data.jl [hunk 5]
```diff
@@ -136,17 +136,124 @@ Tables.materializer(D::Type{<:Data}) = D
...
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, var::Symbol)
+  if var == :geometry
+    view(domain(data), inds)
+  else
+    table  = values(data)
+    subset = Tables.subset(table, inds)
+    cols   = Tables.columns(subset)
+    Tables.getcolumn(cols, var)
+  end
```
Comment (juliohm) [inline src/traits/data.jl]: I think we can also leverage the view(data, inds) here to simplify the code. After we have the view it is easy to use domain(view) and getproperty(view, var) to build a new table.

File: src/traits/data.jl [hunk 6]
```diff
@@ -136,17 +136,124 @@ Tables.materializer(D::Type{<:Data}) = D
...
+  getindex(data, inds, Symbol(var))
+
+function Base.getindex(data::Data, ::Colon, vars::AbstractVector{Symbol})
+  _checkvars(vars)
+  _rmgeometry!(vars)
+  dom   = domain(data)
+  table = values(data)
+
+  cols = Tables.columns(table)
+  𝒯 = NamedTuple(var => Tables.getcolumn(cols, var) for var in vars)
```
Comment (juliohm) [inline src/traits/data.jl]: let's rename T by pairs and use (; pairs...) instead of explicitly writing NamedTuple

File: src/traits/data.jl [hunk 7]
```diff
@@ -136,17 +136,113 @@ Tables.materializer(D::Type{<:Data}) = D
...
+Base.getindex(data::Data, ind::Int, vars::AbstractVector{<:AbstractString}) =
+  getindex(data, ind, Symbol.(vars))
+
+function Base.getindex(data::Data, ind::Int, ::Colon)
+  dom   = domain(data)
+  table = values(data)
+  row   = Tables.subset(table, ind)
+  vars  = Tables.columnnames(row)
+  pairs = (var => Tables.getcolumn(row, var) for var in vars)
+  (; pairs..., geometry=dom[ind])
```
Comment (juliohm) [inline src/traits/data.jl]: This one can be written in terms of the one with ind::Int, vars::AbstractVector

File: src/traits/data.jl [hunk 8]
```diff
@@ -136,17 +136,113 @@ Tables.materializer(D::Type{<:Data}) = D
...
+  table = values(data)
+
+  newdom   = view(domain(data), inds)
+  subset   = Tables.subset(table, inds)
+  cols     = Tables.columns(subset)
+  pairs    = (var => Tables.getcolumn(cols, var) for var in vars)
+  newtable = (; pairs...) |> Tables.materializer(table)
+
+  vals = Dict(paramdim(newdom) => newtable)
+  constructor(data)(newdom, vals)
```
Comment (juliohm) [inline src/traits/data.jl]: This one can be written in terms of view

File: src/traits/data.jl [hunk 9]
```diff
@@ -136,17 +136,113 @@ Tables.materializer(D::Type{<:Data}) = D
...
+end
+
+Base.getindex(data::Data, inds::AbstractVector{Int}, vars::AbstractVector{<:AbstractString}) =
+  getindex(data, inds, Symbol.(vars))
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, ::Colon)
+  dataview = view(data, inds)
+  constructor(data)(domain(dataview), values(dataview))
```
Comment (juliohm) [inline src/traits/data.jl]: This one can be written in terms of the one with inds::AbstractVector, vars::AbstractVector

File: src/traits/data.jl [hunk 10]
```diff
@@ -136,17 +136,113 @@ Tables.materializer(D::Type{<:Data}) = D
...
+end
+
+function Base.getindex(data::Data, inds::AbstractVector{Int}, var::Symbol)
+  dataview = view(data, inds)
+  getproperty(dataview, var)
+end
```
Comment (juliohm) [inline src/traits/data.jl]:
```suggestion
Base.getindex(data::Data, inds::AbstractVector{Int}, var::Symbol)
  getproperty(view(data, inds), var)
```

File: src/traits/data.jl [hunk 11]
```diff
@@ -136,17 +136,113 @@ Tables.materializer(D::Type{<:Data}) = D
...
+  table = values(data)
+  row   = Tables.subset(table, ind)
+  pairs = (var => Tables.getcolumn(row, var) for var in vars)
+  (; pairs..., geometry=dom[ind])
+end
+
+Base.getindex(data::Data, ind::Int, vars::AbstractVector{<:AbstractString}) =
+  getindex(data, ind, Symbol.(vars))
+
+function Base.getindex(data::Data, ind::Int, ::Colon)
```
Comment (juliohm) [inline src/traits/data.jl]: Let's move all the Colon methods to the bottom of the list for better organization

File: src/traits/data.jl (hunk 8 above)
Comment (eliascarv) [reply src/traits/data.jl]: In this case, the number of lines would be the same:
```julia
  dataview = view(data, inds)
  newdom   = domain(dataview)
  cols     = Tables.columns(values(dataview))
  pairs    = (var => Tables.getcolumn(cols, var) for var in vars)
  newtable = (; pairs...) |> Tables.materializer(table)

File: src/traits/data.jl (hunk 9 above)
Comment (eliascarv) [reply src/traits/data.jl]: In this case, I don't think it's necessary for the following reasons:
1 - This method has few lines;
2 - The method `Base.getindex(data::Data, inds::AbstractVector{Int}, vars::AbstractVector{Symbol})` performs a larger number of operations.

File: src/traits/data.jl [hunk 12]
```diff
@@ -136,17 +136,111 @@ Tables.materializer(D::Type{<:Data}) = D
...
+  table = values(data)
+
+  newdom   = view(domain(data), inds)
+  subset   = Tables.subset(table, inds)
+  cols     = Tables.columns(subset)
+  pairs    = (var => Tables.getcolumn(cols, var) for var in vars)
+  newtable = (; pairs...) |> Tables.materializer(table)
+
+  vals = Dict(paramdim(newdom) => newtable)
+  constructor(data)(newdom, vals)
```
Comment (juliohm) [inline src/traits/data.jl]: Let's be more systematic here and define dom = domain(data), tab = values(data), then define newdom = view(dom, inds), newtab = ...

File: src/traits/data.jl [hunk 13]
```diff
@@ -136,17 +136,111 @@ Tables.materializer(D::Type{<:Data}) = D
...
+end
+
+Base.getindex(data::Data, inds::AbstractVector{Int}, var::AbstractString) =
+  getindex(data, inds, Symbol(var))
+
+function Base.getindex(data::Data, ind::Int, vars::AbstractVector{Symbol})
+  _checkvars(vars)
+  _rmgeometry!(vars)
+  dom   = domain(data)
+  table = values(data)
```
Comment (juliohm) [inline src/traits/data.jl]:
```suggestion
  tab = values(data)
```

File: src/traits/data.jl [hunk 14]
```diff
@@ -136,17 +136,111 @@ Tables.materializer(D::Type{<:Data}) = D
...
+
+Base.getindex(data::Data, ind::Int, var::Symbol) =
+  getproperty(data, var)[ind]
+
+Base.getindex(data::Data, ind::Int, var::AbstractString) =
+  getindex(data, ind, Symbol(var))
+
+function Base.getindex(data::Data, ind::Int, ::Colon)
+  dom   = domain(data)
+  table = values(data)
```
Comment (juliohm) [inline src/traits/data.jl]:
```suggestion
  tab = values(data)
```

File: src/traits/data.jl [hunk 15]
```diff
@@ -136,17 +136,111 @@ Tables.materializer(D::Type{<:Data}) = D
...
+  vars  = Tables.columnnames(row)
+  pairs = (var => Tables.getcolumn(row, var) for var in vars)
+  (; pairs..., geometry=dom[ind])
+end
+
+function Base.getindex(data::Data, ::Colon, vars::AbstractVector{Symbol})
+  _checkvars(vars)
+  _rmgeometry!(vars)
+  dom   = domain(data)
+  table = values(data)
```
Comment (juliohm) [inline src/traits/data.jl]:
```suggestion
  tab = values(data)
```

File: src/traits/data.jl [hunk 16]
```diff
@@ -136,17 +136,111 @@ Tables.materializer(D::Type{<:Data}) = D
...
+
+  cols     = Tables.columns(table)
+  pairs    = (var => Tables.getcolumn(cols, var) for var in vars)
+  newtable = (; pairs...) |> Tables.materializer(table)
```
Comment (juliohm) [inline src/traits/data.jl]:
```suggestion
  newtab = (; pairs...) |> Tables.materializer(table)
```

File: src/traits/data.jl [hunk 17]
```diff
@@ -136,17 +136,111 @@ Tables.materializer(D::Type{<:Data}) = D
...
+
+  vals = Dict(paramdim(dom) => newtable)
```
Comment (juliohm) [inline src/traits/data.jl]:
```suggestion
  newvals = Dict(paramdim(dom) => newtable)
```

File: src/traits/data.jl [hunk 18]
```diff
@@ -136,17 +136,106 @@ Tables.materializer(D::Type{<:Data}) = D
...
+function Base.getindex(data::Data, ind::Int, ::Colon)
+  dom   = domain(data)
+  tab   = values(data)
+  row   = Tables.subset(tab, ind)
+  vars  = Tables.columnnames(row)
+  pairs = (var => Tables.getcolumn(row, var) for var in vars)
+  (; pairs..., geometry=dom[ind])
+end
+
+function Base.getindex(data::Data,::Colon, vars::AbstractVector{Symbol})
```
Comment (eliascarv) [inline src/traits/data.jl]:
```suggestion
function Base.getindex(data::Data, ::Colon, vars::AbstractVector{Symbol})
```

**Pull request merged**